File: Difficulty: Easy/Largest number in one swap/largest-number-in-one-swap.cpp
class Solution {
  public:
    string largestSwap(string &s) {
        // code here
        unordered_map<int,int>mp;
        int n=s.size();
        for(int i=n-1;i>=0;i--)
        {
            if(mp.find(s[i]-'0')==mp.end())
            {
                mp[s[i]-'0']=i;
            }
        }
        for(int i=0;i<n;i++)
        {
            int x=s[i]-'0';
            for(int j=9;j>x;j--)
            {
                if(mp[j]>i)
                {
                  swap(s[i],s[mp[j]]);
                  return s;
                }
            }
        }
        return s;
        
    }
};

File: Difficulty: Easy/Largest number in one swap/largest-number-in-one-swap.java
class Solution {
    public String largestSwap(String s) {
        // code here
          HashMap<Integer, Integer> mp = new HashMap<>();
        int n = s.length();

        for (int i = n - 1; i >= 0; i--) {
            int digit = s.charAt(i) - '0';
            if (!mp.containsKey(digit)) {
                mp.put(digit, i);
            }
        }

        char[] arr = s.toCharArray();

        for (int i = 0; i < n; i++) {
            int x = arr[i] - '0';

            for (int j = 9; j > x; j--) {
                if (mp.containsKey(j) && mp.get(j) > i) {
                    int idx = mp.get(j);

                    char temp = arr[i];
                    arr[i] = arr[idx];
                    arr[idx] = temp;

                    return new String(arr);
                }
            }
        }

        return s;
    }
}